reactos/drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2752 @@
+
+        TdVA++;
+        TdPA++;
+    }
+}
+
+MPSTATUS
+NTAPI
+OHCI_OpenControlEndpoint(IN POHCI_EXTENSION OhciExtension,
+                         IN PUSBPORT_ENDPOINT_PROPERTIES EndpointProperties,
+                         IN POHCI_ENDPOINT OhciEndpoint)
+{
+    POHCI_HCD_ED ED;
+
+    DPRINT_OHCI("OHCI_OpenControlEndpoint: ... \n");
+
+    ED = (POHCI_HCD_ED)EndpointProperties->BufferVA;
+
+    OhciEndpoint->FirstTD = (POHCI_HCD_TD)((ULONG_PTR)ED + sizeof(OHCI_HCD_ED));
+    OhciEndpoint->HeadED = &OhciExtension->ControlStaticED;
+
+    OHCI_InitializeTDs(OhciEndpoint, EndpointProperties);
+
+    OhciEndpoint->HcdED = OHCI_InitializeED(OhciEndpoint,
+                                            ED,
+                                            OhciEndpoint->FirstTD,
+                                            EndpointProperties->BufferPA);
+
+    OhciEndpoint->HcdED->Flags = OHCI_HCD_ED_FLAG_CONTROLL |
+                                 OHCI_HCD_ED_FLAG_RESET_ON_HALT;
+
+    OHCI_InsertEndpointInSchedule(OhciEndpoint);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+OHCI_OpenBulkEndpoint(IN POHCI_EXTENSION OhciExtension,
+                      IN PUSBPORT_ENDPOINT_PROPERTIES EndpointProperties,
+                      IN POHCI_ENDPOINT OhciEndpoint)
+{
+    POHCI_HCD_ED ED;
+
+    DPRINT_OHCI("OHCI_OpenBulkEndpoint: ... \n");
+
+    ED = (POHCI_HCD_ED)EndpointProperties->BufferVA;
+
+    OhciEndpoint->FirstTD = (POHCI_HCD_TD)((ULONG_PTR)ED + sizeof(OHCI_HCD_ED));
+    OhciEndpoint->HeadED = &OhciExtension->BulkStaticED;
+
+    OHCI_InitializeTDs(OhciEndpoint, EndpointProperties);
+
+    OhciEndpoint->HcdED = OHCI_InitializeED(OhciEndpoint,
+                                            ED,
+                                            OhciEndpoint->FirstTD,
+                                            EndpointProperties->BufferPA);
+
+    OHCI_InsertEndpointInSchedule(OhciEndpoint);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+OHCI_OpenInterruptEndpoint(IN POHCI_EXTENSION OhciExtension,
+                           IN PUSBPORT_ENDPOINT_PROPERTIES EndpointProperties,
+                           IN POHCI_ENDPOINT OhciEndpoint)
+{
+    UCHAR Index[8];
+    UCHAR Period;
+    ULONG PeriodIdx = 0;
+    POHCI_HCD_ED ED;
+    ULONG ScheduleOffset;
+
+    DPRINT_OHCI("OHCI_OpenInterruptEndpoint: ... \n");
+
+    ED = (POHCI_HCD_ED)EndpointProperties->BufferVA;
+
+    OhciEndpoint->FirstTD = (POHCI_HCD_TD)((ULONG_PTR)ED + sizeof(OHCI_HCD_ED));
+
+    Index[0] = ENDPOINT_INTERRUPT_1ms - 1;
+    Index[1] = ENDPOINT_INTERRUPT_2ms - 1;
+    Index[2] = ENDPOINT_INTERRUPT_4ms - 1;
+    Index[3] = ENDPOINT_INTERRUPT_8ms - 1;
+    Index[4] = ENDPOINT_INTERRUPT_16ms - 1;
+    Index[5] = ENDPOINT_INTERRUPT_32ms - 1;
+    Index[6] = ENDPOINT_INTERRUPT_32ms - 1;
+    Index[7] = ENDPOINT_INTERRUPT_32ms - 1;
+
+    Period = EndpointProperties->Period;
+
+    while (!(Period & 1))
+    {
+        PeriodIdx++;
+        Period >>= 1;
+    }
+
```

Comment:
Assert (or check) that PeriodIdx < arraysize(Index)

Reply:
Ok.